File: files/2024-10-27/CSEC 27.10.py
import pandas as pd

path = "files/Excel Challenge October 27th.xlsx"
input = pd.read_excel(path, usecols="C:E", skiprows=2, nrows=19)
test = pd.read_excel(path, usecols="G:H", skiprows=2, nrows=5).dropna().rename(columns=lambda x: x.replace('.1', ''))
test['Date'] = pd.to_datetime(test['Date'])

threshold = 800

result = (input[input.groupby('Staff')['Sales'].cumsum() >= threshold]
          .drop_duplicates('Staff')
          .assign(Date=lambda df: pd.to_datetime(df['Date']))
          .loc[:, ['Staff', 'Date']]
          .sort_values(by='Date')
          .reset_index(drop=True))

print(result.equals(test)) # True

File: files/2024-10-27/CSEC 27.10.R
library(tidyverse)
library(readxl)

path = "files/Excel Challenge October 27th.xlsx"
input = read_excel(path, range = "C3:E21")
test  = read_excel(path, range = "G3:H7") %>% na.omit() %>% mutate(Date = as.Date(Date))
threshold = 800

result = input %>%
  group_by(Staff) %>%
  filter(cumsum(Sales) >= threshold) %>%
  slice(1) %>%
  mutate(Date = as.Date(Date)) %>%
  select(Staff, Date) %>%
  arrange(Date)

all.equal(result, test, check.attributes = FALSE)

